### PR TITLE
feat: add `add` and `remove` methods to modify the playlist dynamically

### PR DESCRIPTION
**Pull request by misteroneill**
## Description
This was a feature request from a Brightcove customer: the ability to add or remove single items from the playlist without replacing the entire playlist. Seemed like an easy update and one that would be sensible to make!

## Specific Changes proposed
* New `add` function for adding playlist items
* New `remove` function for removing playlist items
* Simplified logic for preparing playlist items so that the `add` and `remove` methods could be added with minimal code duplication
* Updated some metafiles

## Requirements Checklist
- [x] Feature implemented / Bug fixed
- [x] If necessary, more likely in a feature request than a bug fix
  - [x] Unit Tests updated or fixed
  - [x] Docs/guides updated
- [ ] Reviewed by Two Core Contributors

### CODE REVIEW

File: src/playlist-maker.js [hunk 1]
```diff
@@ -236,28 +222,17 @@ export default function factory(player, initialList, initialIndex = 0) {
...
-      // Transform any primitive and null values in an input list to objects
-      if (list.filter(item => isItemObject(item)).length !== list.length) {
-        list = transformPrimitiveItems(list);
-      }
-
-      // Add unique id to each playlist item. This id will be used
-      // to determine index in cases where there are more than one
-      // identical sources in the playlist.
-      generatePlaylistItemId(list);
+      list = preparePlaylistItems(nextPlaylist);
```
Comment (misteroneill) [inline src/playlist-maker.js:235]: Logic in this function significantly simplified.

File: src/playlist-maker.js [hunk 2]
```diff
@@ -372,6 +347,55 @@ export default function factory(player, initialList, initialIndex = 0) {
...
+    if (changing) {
+      throw new Error('removeItem failed: playlist is currently changing');
+    }
+    if (index >= 0 && index < list.length) {
+      list.splice(index, 1);
+      player.trigger('playlistchange');
+      player.trigger({type: 'playlistitemremoved', index});
+    }
+  };
+
```
Comment (misteroneill) [inline src/playlist-maker.js:398]: New methods added here.

File: src/playlist-maker.js [hunk 3]
```diff
@@ -711,7 +735,7 @@ export default function factory(player, initialList, initialIndex = 0) {
 
   // If an initial list was given, populate the playlist with it.
   if (Array.isArray(initialList)) {
-    playlist(initialList.slice(), initialIndex);
+    playlist(initialList, initialIndex);
```
Comment (misteroneill) [inline src/playlist-maker.js:738]: This `slice()` was redundant. How many times do we need to shallow-clone the array?!

**Review by gjanblaszczyk — APPROVED** [review]
LGTM!

**Review by mister-ben — COMMENTED** [review]
Other than the docs comment looks good. However since the obvious next request will be to add/insert multiple items in this way I think it would make sense to support that too.

File: docs/api.md [hunk 4]
```diff
@@ -111,61 +111,84 @@ player.playlist.currentItem();
 // -1
 ```
 
-#### `player.playlist.addItem(String|Object newItem, [Number index])`
+#### `player.playlist.add(String|Object|Array items, [Number index])`
 
 Adds an item to the current playlist without replacing the playlist.
```
Comment (mister-ben) [inline docs/api.md:116]:
```suggestion
Adds one or more items to the current playlist without replacing the playlist.
```

File: docs/api.md [hunk 5]
```diff
@@ -111,61 +111,84 @@ player.playlist.currentItem();
...
 
-Fires the `playlistitemadded` event.
+Fires the `playlistadded` event.
```
Comment (mister-ben) [inline docs/api.md:118]: We should mention that `playlistchange` is not fired in this doc too, since it could easily be assumed otherwise.

**Review by mister-ben — COMMENTED** [review]
The lack of `playlistchange` event could cause some confusion. It is currently described later in this file as "fired ... whenever the contents of the playlist are changed", and `add()`/`remove()` do change the contents. We'll need to update that descripition too.

File: docs/api.md (hunk 5 above)
Comment (gkatsev) [reply docs/api.md:118]: should playlistchange fire for added/removed events too?

Comment (gkatsev) [reply docs/api.md:118]: also, I think this is the `playlistadd` event, not `playlistadded`
https://github.com/videojs/videojs-playlist/blob/50503ca20bb935652ac785b04f9f0d549f975401/src/playlist-maker.js#L410

Comment (misteroneill) [reply docs/api.md:118]: Yeah, I was going back and forth on `playlistchange` but I will restore it.

**Review by mister-ben — APPROVED** [review]

**Pull request merged**